external_apps/bid/bid_action/counter_offer.go

```diff
@@ -0,0 +1,194 @@
+}
+
+func runCounterOffer(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	counterOffer := CounterOffer{}
+	err := counterOffer.Unmarshal(tx.Data)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, action.ErrWrongTxType, counterOffer.Tags(), err)
+	}
+
+	//1. verify bidConvId exists in ACTIVE store
+	bidMasterStore, err := GetBidMasterStore(ctx)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrGettingBidMasterStore, counterOffer.Tags(), err)
+	}
+
+	if !bidMasterStore.BidConv.WithPrefixType(bid_data.BidStateActive).Exists(counterOffer.BidConvId) {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrBidConvNotFound, counterOffer.Tags(), err)
+	}
+
+	bidConv, err := bidMasterStore.BidConv.WithPrefixType(bid_data.BidStateActive).Get(counterOffer.BidConvId)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrGettingBidConv, counterOffer.Tags(), err)
+	}
+
+	//2. check owner's identity
+	if !counterOffer.AssetOwner.Equal(bidConv.AssetOwner) {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrWrongAssetOwner, counterOffer.Tags(), err)
+	}
+
+	//3. check expiry
+	deadLine := time.Unix(bidConv.DeadlineUTC, 0)
+
+	if deadLine.Before(ctx.Header.Time.UTC()) {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrExpiredBid, counterOffer.Tags(), err)
+	}
+
+	//4. check asset availability
+	available, err := IsAssetAvailable(ctx, bidConv.AssetName, bidConv.AssetType, bidConv.AssetOwner)
+	if err != nil || available == false {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrInvalidAsset, counterOffer.Tags(), err)
+	}
+
+	//5. get active bid offer
+	activeOffer, err := bidMasterStore.BidOffer.GetActiveOffer(counterOffer.BidConvId, bid_data.TypeBidOffer)
+	if err != nil || activeOffer == nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrGettingActiveBidOffer, counterOffer.Tags(), err)
+	}
+	//6. amount needs to be large than active bid offer from bidder
+	offerCoin := counterOffer.Amount.ToCoin(ctx.Currencies)
+	activeOfferCoin := activeOffer.Amount.ToCoin(ctx.Currencies)
+	if offerCoin.LessThanEqualCoin(activeOfferCoin) {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrAmountLessThanActiveOffer, counterOffer.Tags(), err)
+	}
+
+	//7. unlock bidder's previous amount and deactivate the bidder's offer
+	// this way we only lock amount from a bid offer from bidder
+	// if the active offer is a counter offer from owner, no amount is locked from the bidder
+
+	err = DeactivateOffer(false, bidConv.Bidder, ctx, activeOffer, bidMasterStore)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrDeactivateOffer, counterOffer.Tags(), err)
+	}
+
+	//8. add new counter offer to offer store
+	createCounterOffer := bid_data.NewBidOffer(
+		counterOffer.BidConvId,
+		bid_data.TypeCounterOffer,
+		ctx.Header.Time.UTC().Unix(),
+		counterOffer.Amount,
+		bid_data.CounterOfferAmount,
+	)
+
```

Comment:
Do we need to make make sure the conversation does not expire in this exact block ?

Reply:
yeah there is in line 110

Reply:
ok you are talking about expire in exact this block.
then I think we are fine, because in next block it will be expired

Reply:
Yeah , but the bidder wont have a chance to agree or disagree to it ?

Reply:
yeah, but if we allow one or two blocks for the other party to respond, then the deadline showing in bidConv will be no longer accurate.

Reply:
I was thinking ,maybe we can increment the expiry everytime a counter offer is made .by a fixed amount .

Reply:
yeah good idea, or maybe we can just increment the expiry after any new changes